### PR TITLE
Ins 40 instance domain

### PR DESCRIPTION
**Pull request by ilyapt**
**- What I did**
Instance Domain
Instance Domain Proxy
Instance Domain Factory
Tests

**- How to verify it**
Run tests

### CODE REVIEW

File: genesis/public/core/instance.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+// InstanceDomainName is a name for instance domain.
+const InstanceDomainName = "InstanceDomain"
+
+// InstanceDomain is a contract that stores instances of other domains
+type InstanceDomain interface {
+	// Base domain implementation.
+	domain.Domain
+	// CreateInstance is used to new instance as a child to domain storage.
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: You probably forgot word "create" here :) ("is used to **create** new instance")

File: genesis/public/core/instance_test.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+	factory := NewInstanceDomainFactory()
+	proxy := factory.Create(parent)
+	instDomain, _ := newInstanceDomain(parent)
+
+	assert.Equal(t, &instanceDomainProxy{
+		instance: instDomain,
+	}, proxy)
+}
+
+func TestInstanceDomainFactory_CreateWithError(t *testing.T) {
```
Comment (LinnikD) [inline genesis/public/core/instance_test.go]: There need to be two tests with errors:  
- with no parent
- with parent, whose AddChild returns error

**Comment by LinnikD** [comment]
Ok

File: genesis/public/core/instance.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+func (idf *instanceDomainFactory) GetClassID() string {
+	return class.InstanceDomainID
+}
+
+// GetReference returns nil for not published factory
+func (idf *instanceDomainFactory) GetReference() *object.Reference {
+	return nil
+}
+
+// Create factory method is used to create new InstanceDomain instances.
```
Comment (dmitry-panchenko) [inline genesis/public/core/instance.go]: `Create is factory method that used...`
or
`Create - factory method...`

**Pull request merged**